skyd/cache.go

```diff
@@ -0,0 +1,168 @@
+package skyd
+
+import (
+	"sync"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/node/api"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+)
+
+type (
+	// PinnedSkylinksCache is a simple cache of the renter's directory
+	// information, so we don't need to fetch that for each skylink we
+	// potentially want to pin/unpin.
+	PinnedSkylinksCache struct {
+		result   *RebuildCacheResult
+		skylinks map[string]struct{}
+		mu       sync.Mutex
+	}
+	// RebuildCacheResult informs the caller on the status of a cache rebuild.
+	// The error should not be read before the channel is closed.
+	RebuildCacheResult struct {
+		// Ch indicates the status of the cache rebuild progress - if it's not
+		// closed then the rebuild is still in progress.
+		Ch chan struct{}
+		// ExternErr holds the error state of the cache rebuild process. It must
+		// only be read after Ch is closed.
+		ExternErr error
+	}
+	// NodeSkydClient describes the part of skydclient.Client which we need.
+	// This type exists for testing purposes.
+	NodeSkydClient interface {
```

Comment:
I don't really like that we are making multiple skyd client interfaces throughout the code base.  If the reason for doing it here is to simply save a few lines of test code I'd prefer we didn't. 

Just my 2c, if I'm alone on this then that is ok.

Reply:
That's fair. To be honest, initially I thought those were two separate clients. 😅
Then I kept it because of some circular dependencies which forced me to put the mock next to the client and we typically don't like that at all. But in this case it makes some sense. I just went over it with Chris and we stared at it for 10 minutes trying to figure out a better approach and this seems to be the best compromise.

So, TL;DR:
 - I'll merge the two mocks
 - the common mock will be used by both the unit tests and the integration tests
 - the common mock will live in the main package next to the client implementation